alert/alert.go

```diff
@@ -57,12 +59,21 @@ func (a Alert) Normalize() (*Alert, error) {
 	}
 	a.Summary = strings.Replace(a.Summary, "\n", " ", -1)
 	a.Summary = strings.Replace(a.Summary, "  ", " ", -1)
+
+	var validateMeta error
+	for k := range a.Meta.AlertMetaV1 {
+		if k == "" {
+			validateMeta = validation.NewFieldError("Meta", "key must be non empty string")
+		}
+	}
```

Comment:
If we use a separate method, we can handle validation there.

That aside, we should set sane limits on the number of keys and key & value lengths also. We should also probably limit the keys to ASCII values or some sort of subset.

Lastly, just a nit, but we should break the loop if we find this.

Reply:
Okay, Is there any convention followed in limiting these values. I saw that `label Key has limit of 255 in the codebase`, should the meta `Key` in meta also follow the same?

Reply:
Yeah, I think that's reasonable for the key -- it's possible we may index the keys in the future, so why don't we go with:


- keys: validate.ASCII 1-255 characters
- values: just limit the length to 2k/2048 for now, (see note below)
- total number of key/value pairs: let's start with 16 (you can do `len()` on the map) (see note below)

note:
For values and # of keys, we don't have validation functions for those. For values we don't want to use `validate.Text` because it restricts whitespace and we want to preserve the integrity/exactness of metadata.

Here's what that may look like and how to create a validation error for the value

```go
if len(meta) > 16 { // meta is of type map[string]string
	return validation.NewFieldError("Meta", "must not have more than 16 keys")
}

var err error
for k, v :=range meta {
	fieldName := fmt.Sprintf("Meta[%s]", k)
	err = validate.Many(err, validate.ASCII(fieldName, k))
	if len(v) > 2048 {
		err = validate.Many(err,
			validation.NewFieldError(fieldName, "cannot be more than 2048 characters"),
		)
	}
}
if err != nil {
	return err
}
```

Reply:
One thing I'm wondering is if we should truncate long values, rather than rejecting. Keys can be pre-defined, but values may likely be a pass-through from other systems.

Since value may be arbitrary, it'd likely be better to truncate and create an alert, than have certain alerts fail to create (i.e., due to unexpected data)

Reply:
sounds good!